### PR TITLE
Allow enabling/disabling Incremental Prebuilds in Project Settings

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

Allow enabling/disabling Incremental Prebuilds in Project Settings:

- [x] Add a new Project Settings menu containing:
  - [x] The existing Configuration page
  - [x] Another page with a "Enable Incremental Prebuilds" checkbox
- [x] If it's enabled for a Project, make new Prebuilds incremental
- [x] In the Prebuild page, show when a Prebuild is incremental
- [x] Drive-by: Improve formatting of Prebuild duration
- [x] Drive-by: Correctly show 'timed out' status in Prebuild Logs

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/6727

## How to test

1. Add a Project
2. Start a Prebuild, and wait for it to be successful
3. Only after it finished successfully, push a new commit to the same branch
4. This should trigger a new Prebuild, but the new Prebuild should not be incremental
5. Next, in the Project's Settings, enable "Use Incremental Prebuilds"
6. Push another commit to the same branch
7. The newly triggered Prebuild should be incremental

## Release Notes

```release-note
Allow enabling/disabling Incremental Prebuilds in Project Settings
Also improve Prebuild Logs UX (show incremental status, format duration nicely, correctly show 'timed out' status)
```

## Documentation

/uncc

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

Approval requirements bypassed by manually added approval.

This pull-request has been approved by:

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

Approval requirements bypassed by manually added approval.

This pull-request has been approved by:

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/projects/Prebuild.tsx
Comment (jankeromnes) [inline components/dashboard/src/projects/Prebuild.tsx:20]: Self-nit: No longer necessary since this rule has been disabled at the root.

File: components/dashboard/src/projects/Project.tsx
Comment (jankeromnes) [inline components/dashboard/src/projects/Project.tsx:22]: Self-nit: Same, remove.

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (jankeromnes) [inline components/dashboard/src/projects/ProjectSettings.tsx:15]: Self-nit: Same, remove.

**Comment by jankeromnes** [comment]
@gtsiolis Please take a look at the UX when you have time 🙏

@geropl Please take a look at the code change when you have time 🙏

**Comment by geropl** [comment]
Starting review now...

**Comment by geropl** [comment]
First impressions: This is set-up for confusion (Settings vs Configuration):
image
I had to look three times to understand where I am and to grok what the difference is.

**Comment by geropl** [comment]
@jankeromnes  Whoops. What did I break? :grimacing:  No prebuild ran yet.
image

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (geropl) [inline components/dashboard/src/projects/Prebuilds.tsx:260]: In the past we used [moment](https://github.com/gitpod-io/gitpod/blob/7abe78c91792c18b0c22103c890b25ea3a4cdf05/components/dashboard/package.json#L10) for this kind of work. Why not re-use that?

File: components/gitpod-db/src/typeorm/entity/db-project.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/entity/db-project.ts:47]: :+1:

File: components/gitpod-db/src/typeorm/migration/1638025268018-ProjectSettings.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/migration/1638025268018-ProjectSettings.ts:14]: :+1:

File: components/gitpod-db/src/typeorm/project-db-impl.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/project-db-impl.ts:73]: For these kind of methods we usually have something like `updatePartial(shape: Partial<Project> & Pick<Project, "id">)` to avoid the repetition. Having the not-exists-check in here is a good addition!

File: components/server/src/projects/projects-service.ts
Comment (geropl) [inline components/server/src/projects/projects-service.ts:241]: This is equvalent to
```
const settings = {
    ...(project.settings || {}),
    ...partial,
}
```
, no?
Also, we check for "project does not exist" here, again.

This is actually the 3rd time:
 - here
 - in our DB abstraction
 - typeorms `save(...)` method
Ideally we'd reduce that to one. If we rely on this in our code we should push it to the DB. And use an `update(...)` instead, which saves us the 3rd check.

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [inline components/server/src/workspace/gitpod-server-impl.ts:1604]: As mentioned above: ideally we'd have a single `udpateProjectPartial` to reduce API implementation.

**Comment by geropl** [comment]
@jankeromnes  Finished 1st round! Nice to see this landing. :flight_arrival: 

I have no hard feelings about the UX "settings vs config". But if 1/2 hour of discussion might solve the (potential) confusion now (maybe with @gtsiolis ?) I feel that'd be well invested time.

**Comment by geropl** [comment]
> I was sort of assuming that the Configuration page might be completely removed soon (e.g. superseded by #6921), or that we could at least move it into Settings

Ah, excellent points! :bulb:

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:72]: **suggestion**: What do you think of using the same beta label as used in the Desktop IDE settings?

https://github.com/gitpod-io/gitpod/blob/c7c0aca070105d894643452a92bd6f79eab44a7f/components/dashboard/src/settings/Preferences.tsx#L109-L113

| Incremental Prebuilds | Desktop IDE |
|-|-|
| Screenshot 2021-12-07 at 11 00 56 AM | Screenshot 2021-12-07 at 11 01 09 AM |

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:71]: **praise**: Thanks @nisarhassan12 for adding the individual changelog entry pages[[1](https://github.com/gitpod-io/website/pull/989)][[2](https://github.com/gitpod-io/website/issues/681)][[3](https://github.com/gitpod-io/website/issues/981)]!

**Review by gtsiolis — APPROVED** [review]
Thanks, @jankeromnes!

Natural next steps could be to introduce an alert component that nudges users to enable or disable incremental prebuilds based on average prebuild duration maybe of the last few prebuilds. 💭 

Approving to unblock merging but holding in case we could resolve the review comments below. Let me know if I missed anything worth discussing here or in sync.

/hold

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: c5ad5c121ae46a9e61b4e3e13eb219fe2bb4d75d</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7031#pullrequestreview-824880786" title="Approved">gtsiolis</a>*

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/Prebuilds.tsx:260]: Hmm, `moment` doesn't seem that great for _formatting time differences_ specifically. (Also, their [docs](https://momentjs.com/docs/) suggest maybe not using it anymore in general.)

The primary advantage of this custom code is doing `mm:ss` when duration is less than 1 hour (e.g. `03:24`), and `h:mm:ss` when it's longer (e.g. `1:05:12`).

I guess we can achieve something similar with `moment` (using two different formats depending on >= or < 1 hour):

```ts
function formatDuration(milliseconds) {
    if (milliseconds < 1000 * 60 * 60) {
        return moment(milliseconds).format('mm:ss');
    }
    return moment(milliseconds).format('hh:mm:ss');
}
```

However (contrived edge case alert 🙈) the above code doesn't properly handle durations longer than 24 hours.

For example, for 25 hours (`milliseconds = 1000 * 60 * 60 * 25`):
- the `moment` code will return `01:00:00` (wrong!)
- the custom code will return `25:00:00` (still correct)

I guess we could handle this with yet another condition, but at this point the complexity with `moment` becomes similar to without `moment` 😋 (or, we just ignore the edge case, which is 99.99% okay but feels bad somehow).

Comment (jankeromnes) [reply components/dashboard/src/projects/Prebuilds.tsx:260]: Maybe like this?

```ts
function formatDuration(milliseconds) {
    const hours = Math.floor(milliseconds / (1000 * 60 * 60));
    return (hours > 0 ? `${hours}:` : '') + moment(milliseconds).format('mm:ss');
}
```

This covers all the edge cases, is short, and uses `moment`. 👍

File: components/dashboard/src/Menu.tsx
Comment (geropl) [reply components/dashboard/src/Menu.tsx:109]: > I'd expect this to always be ordered last

Agree :100:

**Comment by jankeromnes** [comment]
Many thanks! Already addressed most comments, more to follow.

New work-in-progress Project Settings menu:

| Project Configuration | Project Settings |
| --- | --- |
| Screenshot 2021-12-08 at 09 50 53 | Screenshot 2021-12-08 at 09 51 09 |

File: components/gitpod-db/src/typeorm/project-db-impl.ts
Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/project-db-impl.ts:73]: Sure, thanks for the hint! Went with a `DeepPartial` (so that properties like `settings` can also be Partial themselves).

**Comment by jankeromnes** [comment]
Many thanks for the reviews! 🙏 I've addressed all your feedback I think:

- Moved the Configuration page into the new Project Settings, and added a Project Settings side-menu
- Refactored `updateProjectSettings` to a more generic `updateProjectPartial`
- Re-styled the Checkbox
- Fixed a few more bugs

@geropl Could you please take another look at the code when you have time? 🙏

@gtsiolis Could you please check the UX again, in particular the new Project Settings menu? 🙏

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:87]: **issue**: Could we use `max-w-2xl` for the checbox text here to limit the width of this element as we do with similar sections like Account settings?

**fyi**: The grid layout from earlier mockups is not accurately translated in the dashboard yet but will become more relevant once we pick up issues related to responsive design of the dashboard. Posting below an early draft of the responsive design of the grid. Using `max-w-2xl` here is a workaround to approach as much as possible the designs we've been using so far to create visual balance when needed. Feedback is welcome!

| Grid layout in large, medium, small and extra small viewports |
|-|
| Screenshot 2021-12-09 at 10 27 52 PM |

This is also an issue with other parts of the product, see [relevant discussion](https://gitpod.slack.com/archives/C02CBKZEVS8/p1638795733242600?thread_ts=1638366112.229500&cid=C02CBKZEVS8) (internal). Cc @gitpod-io/engineering-ide

File: components/dashboard/src/teams/TeamSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/teams/TeamSettings.tsx:58]: **suggestion**: What do you think of :a: renaming this menu _General_ to follow the pattern we use for team settings and :b: reordering this so it's always the first sub menu entry?

File: components/dashboard/src/teams/TeamSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/teams/TeamSettings.tsx:60]: **issue**: This one also needs to be limited in width with `max-w-2xl`

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:234]: **question**: I could be missing something but do we think it makes sense to remove the latest prebuild block here and only keep the configuration n now that we've removed this from the onboarding flow? Later on we could also handle cases where no configuration is added yet.

**thought**: Eventually we will need to separate project addition and project configuration steps to that these can be done separastely.

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:85]: **suggestion**: We already surface the project name in the top navigation menu. Shallow we remove it from here?

```suggestion
    return <PageWithSubMenu subMenu={getProjectSettingsMenu(project, team)} title="Settings" subtitle="Manage project settings and configuration">
```

**Review by gtsiolis — APPROVED** [review]
Great changes, @jankeromnes! 🧙 

Excited to see this feature becoming part of project settings! 🙌 

Approving to unblock merging but holding because of some comments below and before @geropl take a closer look at the code changes. 🏀  

/hold

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: b0beeec6cfe09f3bfe3f800bf346c69055dd3f28</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7031#pullrequestreview-828154934" title="Approved">gtsiolis</a>*

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by gtsiolis** [comment]
**praise**: @jankeromnes I wanted to drop by again and say thanks again for always using the PR description for breaking down the changes of each PR and including clear steps to test a PR. 🌟

File: components/gitpod-db/src/typeorm/project-db-impl.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/project-db-impl.ts:69]: :+1:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [inline components/server/src/workspace/gitpod-server-impl.ts:1725]: :100: Thx! IMO we should far often be more careful like this, especially when adding super-powerful methods that might change over time. :slightly_smiling_face:

File: components/gitpod-protocol/src/teams-projects-protocol.ts
Comment (geropl) [inline components/gitpod-protocol/src/teams-projects-protocol.ts:64]: :+1:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [inline components/server/src/workspace/gitpod-server-impl.ts:1721]: nit: let's add a `traceAPIParams` like so
```suggestion
        traceAPIParams(ctx, { partialProject: { id: partialProject.id, settings: partialProject.settings });   // censor everything irrelevant

        const user = this.checkUser("updateProjectPartial");
```

**Comment by geropl** [comment]
/lgtm

@jankeromnes I keep the `hold` in case you want to tackle https://github.com/gitpod-io/gitpod/pull/7031#discussion_r766471777 within this PR.

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7031#issuecomment-990735262" title="LGTM">geropl</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7031#pullrequestreview-828154934" title="Approved">gtsiolis</a>*

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl,gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/teams/TeamSettings.tsx
Comment (jankeromnes) [reply components/dashboard/src/teams/TeamSettings.tsx:58]: 🅱️: Sure 👍 happy to make this the top-level entry in the left submenu.

🅰️: I actually find it confusing when the BIG TITLE at the top of the page is different from the Tab I just clicked on: When I click on "Settings", I expect to see "Settings" (and not "General" -- I don't know what that is yet, so I think I mis-clicked).

So I had actually renamed the other "General" title in this PR 🙈

| Project Settings | Team Settings |
| --- | --- |
| Screenshot 2021-12-13 at 15 58 45 | Screenshot 2021-12-13 at 15 58 10 |

You'll notice that I've also renamed the left submenu entry to `Project` and `Team` respectively, because I felt that made sense. However, I'm happy to rename both of these to `General` (but keep `Settings` as the BIG TITLE at the top).

What do you think? 🏓

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:234]: Sure, I'm happy to remove the "Prebuild Runner", i.e. the right part of the split screen, and keep only the editor.

However, this feels like a follow-up issue (because I think we also need to fix the copy).

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ProjectSettings.tsx:85]: Good point, we already see the slug at the top (which should actually be the name, now that I think of it).

Will fix both. 😁

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

File: components/dashboard/src/teams/TeamSettings.tsx
Comment (gtsiolis) [reply components/dashboard/src/teams/TeamSettings.tsx:58]: :b: DEAL 🤝 

:a: Valid concern! This pattern was inherited for team settings from the personal account settings where is makes more sense to differentiate multiple options, see [`/settings`](https://gitpod.io/settings). Keeping the _Settings_ title could make sense for now that there're only two options but could be an issue when project and team settings grow over time. For example, see https://github.com/gitpod-io/gitpod/issues/4456, https://github.com/gitpod-io/gitpod/issues/5067, https://github.com/gitpod-io/gitpod/issues/7101, https://github.com/gitpod-io/gitpod/issues/6706#issuecomment-970649837, etc.

**Comment by geropl** [comment]
/lgtm

Code is fine, and tested last time! :+1:

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: ec9da6fb245e4f3595a832f5ac2fc39bfce1d6ec</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7031#issuecomment-993243906" title="LGTM">geropl</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7031#pullrequestreview-828154934" title="Approved">gtsiolis</a>*

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl,gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
